README.md
# plistutils

## About

The `plistutils` library provides a number of convenience functions for dealing with Apple Property List files.
Our goal is to provide a single, comprehensive Python library for dealing with all aspects of Plist parsing.

plistutils is copyright (c) 2017, Stroz Friedberg, an Aon company.

## Components and Usage

### General Plist Parsing

The `PlistParser` class in `plistparser.py` supports automatic, recursive parsing of binary, XML, and JSON plists.
If an embedded plist is located within the structure of a plist file, calling `PlistParser.parse()` will return
the Python data structure representation of that file and all embedded plists.

#### Sample Usage
```
from plistutils.plistparser import InvalidPlistException, PlistParser

fullpath = 'target.plist'
with open(fullpath, rb) as plist_file:
    try:
        data = PlistParser.parse(plist_file)
    except InvalidPlistException:
        logger.warning("Error parsing plist '{}'", fullpath)
    ...
```

### Alias structure parsing

Beginning in macOS System 7, users could create an "alias" file, which was a dynamically updated link to a
target file, similar in nature to LNK files in Windows. macOS would automatically resolve the location of
the file based on a number of stored attributes. Mac OS X made embedding Alias data structures within plists
regular behavior, also for the purpose of dynamically linking to a target. The Alias structure has undergone
several revisions; the most commonly seen at the time of this publishing are versions 2 and 3, depending on
the application and macOS versions.

Alias data structures have been deprecated, but you will often still find them on machines that have been
upgraded or that are using older versions of application software. For example, if you install Microsoft
Office 2011 on macOS 10.13 High Sierra, the plists created by Office will contain Alias version 2 structures,
while Office 2016 plists will contain a mixture of Alias version 3 and Bookmark structures. Additionally,
Alias data is generally not updated in macOS system plists to newer versions or to Bookmark data unless
the target is updated in the plist (by moving or deleting the plist item, such as in
`com.apple.sidebarlists.plist`).

Both versions 2 and 3 begin with a fixed header structure (containing different fields depending on version),
followed by a table of fields. Fields are noted by a numerical identifier, and not all fields need be
present in the table. The table fields are identical in versions 2 and 3.

The `AliasParser` class in `alias.py` provides parsing for Alias versions 2 and 3. `AliasParser.parse()` is
a generator, and yields Python `dict`s containing the parsed fields. Usually this function will only yield
a single record, but Alias data can contain an embedded Alias record (which could, theoretically, contain
another embedded Alias record), in which case it will yield multiple records.

#### Sample Usage
```
from plistutils.alias import AliasParser
from plistutils.plistparser import PlistParser

fullpath = 'target.plist'
with open(fullpath, rb) as plist_file:
    data = PlistParser.parse(plist_file)
    # Iterate parsed plist and pass index/Alias data to parse_bookmark()
    for alias_record in AliasParser.parse(fullpath, idx, data['Alias']):
        print(alias_record)
```

### Bookmark structure parsing

Bookmark structures are the successor to Alias data. Again, they operate similarly to Windows LNK files, in that
they store attributes about a target, and allow the system to resolve that target. The Bookmark format is a much
more extensible structure allowing the creator to use standard data types or to write arbitrary binary data that
the caller is responsible for storing and interpreting. The detail captured in Bookmarks is often much greater
than that in Alias data.

The `BookmarkParser` class in `bookmark.py` parses Bookmark data. Note that `BookmarkParser.parse_bookmark()` is
a generator which yields multiple parsed records, as a single Bookmark can have multiple entries (usually in the
case of a Bookmark referencing a file within a DMG or similar circumstance) and also contain embedded Alias
data (which could, in turn, contain embedded Alias data).

#### Sample Usage
```
from plistutils.bookmark import BookmarkParser
from plistutils.plistparser import PlistParser

fullpath = 'target.plist'
with open(fullpath, rb) as plist_file:
    data = PlistParser.parse(plist_file)
    # Iterate parsed plist and pass index/name/Bookmark data to parse_bookmark()
    for bookmark_record in BookmarkParser.parse_bookmark(fullpath, idx, name, data['Bookmark']):
        print(bookmark_record)
```

### NSKeyedArchiver parsing

Apple occasionally uses the `NSKeyedArchiver` format to store data within a plist. This "encoder" format has existed
since macOS 10.2, and an example can be seen in the typical iChat plist as well as in SFL recent items plists. This
format is meant to be somewhat space-efficient, as multiple objects can point to the same children values.

The `NSKeyedArchiveParser` in `nskeyedarchiver.py` parses most standard data types, and we intend to improve support
for complex types in future releases.

#### Sample Usage
```
from plistutils.nskeyedarchiver import NSKeyedArchiveParser, NSKeyedArchiveException
from plistutils.plistparser import PlistParser

fullpath = 'target.plist'
with open(fullpath, rb) as plist_file:
    data = PlistParser.parse(plist_file)
    if NSKeyedArchiveParser.is_known_nskeyedarchive(data, fullpath):
        nska_parser = NSKeyedArchiveParser(fullpath)
        try:
            unarchived_data = nska_parser.parse_archive(data).get('root', {})
        except NSKeyedArchiveException:
            return
        print(unarchived_data)
```

### utils.py

`utils.py` contains general utility functions, mostly focused on correct parsing and rounding of timestamps.

## Requirements

`plistutils` depends on the biplist library (https://pypi.python.org/pypi/biplist) and is tested on CPython 3.5 64-bit.

## Acknowledgements

Michael Lynn provided fantastic research illuminating data stored in Bookmarks, without which we could not have
created the Bookmark parser (http://michaellynn.github.io/2015/10/24/apples-bookmarkdata-exposed/).

Patrick Olsen wrote a blog post (http://sysforensics.org/2016/08/mac-alias-data-objects/) which spurred us to take a
look at the Alias v3 structure and produce a fully functional parser.

## Other Plist Utility Libraries

Alastair Houghton maintains a Python library which parses and creates Alias version 2 and Bookmark structures
(https://github.com/al45tair/mac_alias). We came across the `mac_alias` repository after the majority of work was
completed on Alias version 3 and Bookmark parsing.

CCL Group publishes a binary plist parser which supports extracting NSKeyedArchive data when it contains common object
types (https://github.com/cclgroupltd/ccl-bplist).


CHANGELOG.md
## 20171222-01 v1.0.0

### New Features

- Breaking Change - altered Alias data field names to make consistent with Bookmark output
- Support embedded Alias data within Alias and Bookmark records
- Support multiple $top items in NSKeyedArchive plists
- Support embedded NSKeyedArchive data within NSKeyedArchive plists
- Added several new supported data types to NSKeyedArchiveParser
- Utils: new case insensitive dict getter for use with SFL/SFL2 files
- Added setup.py to distribution
- Bumped biplist requirement to 1.0.3


## 20171016-01

### New Features

- Initial release


plistutils/utils.py
from collections import namedtuple
from datetime import datetime
import decimal
import struct
from uuid import UUID


UNIX_EPOCH = datetime(1970, 1, 1)
HFS_EPOCH = datetime(1904, 1, 1)
HFS_EPOCH_FROM_UNIX_SHIFT = (UNIX_EPOCH - HFS_EPOCH).total_seconds()
MAC_ABSOLUTE_TIME_EPOCH = datetime(2001, 1, 1)
MAC_ABSOLUTE_TIME_EPOCH_FROM_UNIX_SHIFT = (UNIX_EPOCH - MAC_ABSOLUTE_TIME_EPOCH).total_seconds()


class NamedStruct(struct.Struct):
    def __init__(self, struct_name='NamedStruct', endianness='<', pairs=None):
        if pairs is None:
            pairs = []
        struct_args = [x[1] for x in pairs]
        super().__init__(endianness + ' '.join(struct_args))
        self.structured_data = namedtuple(struct_name, [x[0] for x in pairs if 'x' not in x[1]])

    def parse(self, buf, offset):
        return self.structured_data._make(self.unpack(buf[offset:self.size + offset]))

    def parse_as_dict(self, buf, offset):
        return dict(self.parse(buf, offset)._asdict())


def guid_str_from_bytes(guid_bytes, endian='le'):
    if endian.lower() == 'le':
        return str(UUID(bytes_le=guid_bytes))
    else:
        return str(UUID(bytes=guid_bytes))


def interpret_flags(bitmask, values):
    """
    Args:
        bitmask: flags to check
        values: array of tuples containing (value, description) pairs

    Returns: string containing descriptions of flags

    """
    return ', '.join(desc for num, desc in values if num & bitmask) if bitmask else None


def parse_mac_absolute_time(seconds, resolution=1):
    try:
        return parse_timestamp(seconds, resolution, MAC_ABSOLUTE_TIME_EPOCH_FROM_UNIX_SHIFT) if seconds else None
    except Exception:
        return None


def parse_timestamp(qword, resolution, epoch_shift, mode=decimal.ROUND_HALF_EVEN):
    """
    Generalized function for parsing timestamps

    :param qword: number of time units since the epoch
    :param resolution: number of time units per second
    :param epoch_shift: difference in seconds between UNIX epoch (1970-1-1)
                        and epoch of qword
    :param mode: decimal rounding mode
    :return: datetime.datetime
    """
    # convert qword from given epoch to UNIX epoch
    shifted = qword - epoch_shift * resolution
    # python's datetime.datetime supports microsecond precision
    datetime_resolution = int(1e6)

    # total number of microseconds since UNIX epoch
    # python 3 round() returns int, python 2 round() returns float
    total_microseconds = (decimal.Decimal(shifted * datetime_resolution) / decimal.Decimal(resolution)).quantize(1, mode)

    # convert to datetime
    return datetime.utcfromtimestamp(total_microseconds // datetime_resolution).replace(microsecond=total_microseconds % datetime_resolution)


def case_insensitive_dict_get(d, key, default=None):
    """
    Searches a dict for the first key matching case insensitively. If there is
    an exact match by case for the key, that match is given preference.

    Args:
        d: dict to search
        key: key name to retrieve case-insensitively

    Returns: value or default

    """
    if not key:
        return default
    if key in d:
        return d[key]
    if isinstance(key, str):
        key_lower = key.lower()
        for k in d:
            if k.lower() == key_lower:
                return d[k]
    return default


plistutils/plistparser.py
from io import BytesIO
import json
import logging
import plistlib


import biplist


logger = logging.getLogger(__name__)


class InvalidPlistException(Exception):
    pass


class PlistParser(object):
    class PlistTypes(object):
        not_plist_type = 0
        binary_type = 1
        xml_type = 2
        json_type = 3

    _binary_magic = b'bplist00'
    _xml_magic = b"<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<!DOCTYPE plist PUBLIC \"-//Apple//DTD PLIST 1.0//EN\" \"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">\n<plist version=\"1.0\">\n"
    _xml2_magic = _xml_magic.replace(b'\n', b'\r\n')
    _xml_str = _xml_magic.decode('utf-8')
    _xml2_str = _xml_magic.decode('utf-8')
    _json_magic_1 = b'['
    _json_magic_2 = b'{'

    @classmethod
    def _get_plist_type(cls, file_obj):
        class NestedScope(object):
            buf = b''
            size = 0

        # Ensures len(NestedScope.buf) >= length of target and compares
        def read_and_check(magic):
            # Already read the required number of bytes
            if len(NestedScope.buf) >= len(magic):
                return NestedScope.buf.startswith(magic)
            # Impossible to satisfy the requirement
            if len(NestedScope.buf) + len(magic) > NestedScope.size:
                return False
            NestedScope.buf += file_obj.read(len(magic) - len(NestedScope.buf))
            return NestedScope.buf.startswith(magic)

        file_obj.seek(0, 2)
        NestedScope.size = file_obj.tell()
        file_obj.seek(0)

        if read_and_check(cls._binary_magic):
            return cls.PlistTypes.binary_type
        if read_and_check(cls._xml_magic) or read_and_check(cls._xml2_magic):
            return cls.PlistTypes.xml_type
        if read_and_check(cls._json_magic_1) or read_and_check(cls._json_magic_2):
            return cls.PlistTypes.json_type
        return cls.PlistTypes.not_plist_type

    @classmethod
    def parse(cls, file_obj):
        file_obj.seek(0)
        return cls._parse(BytesIO(file_obj.read()))

    @classmethod
    def _parse(cls, file_obj, plist_type=None):
        def visit(plist):
            if isinstance(plist, (bytes, str)):
                if isinstance(plist, str):
                    if plist.startswith(cls._xml_str) or plist.startswith(cls._xml2_str):
                        plist = plist.encode('utf-8')
                    else:
                        return None

                try:
                    value_flo = BytesIO(plist)
                    value_type = cls._get_plist_type(value_flo)
                    if value_type != cls.PlistTypes.not_plist_type:
                        return cls._parse(value_flo, plist_type=value_type)
                except Exception:
                    # It wasn't a plist
                    return None
            iterators = {list: lambda x: iter(enumerate(x)),
                         dict: lambda x: iter(x.items())}
            for base_type, value in iterators.items():
                if isinstance(plist, base_type):
                    it = value
                    break
            else:
                it = lambda x: []

            for k, v in it(plist):
                visited = visit(v)
                if visited:
                    plist[k] = visited
            return None

        data = cls._read_plist(file_obj, plist_type=plist_type)
        visit(data)
        return data

    @classmethod
    def _read_plist(cls, file_obj, plist_type=None):
        # TODO: Old-Style ASCII property Lists
        # https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/PropertyLists/OldStylePlists/OldStylePLists.html
        # They're like JSON, except not.
        if plist_type is None:
            plist_type = cls._get_plist_type(file_obj)
        try:
            file_obj.seek(0)
            return ({cls.PlistTypes.binary_type: cls._read_binary_plist,
                     cls.PlistTypes.xml_type: cls._read_xml_plist,
                     cls.PlistTypes.json_type: cls._read_json_plist}
                    [plist_type](file_obj))
        except Exception:
            raise

    @classmethod
    def _read_binary_plist(cls, file_obj):
        try:
            return biplist.readPlist(file_obj)
        except biplist.InvalidPlistException as e:
            raise InvalidPlistException(e)

    @classmethod
    def _read_xml_plist(cls, file_obj):
        rtn = plistlib.load(file_obj)
        return rtn

    @classmethod
    def _read_json_plist(cls, file_obj):
        rtn = json.load(file_obj)
        return rtn


plistutils/bookmark.py
from collections import defaultdict
import logging
import struct
from urllib.parse import urljoin


from plistutils.alias import AliasParser
from plistutils.utils import guid_str_from_bytes, interpret_flags, parse_mac_absolute_time


logger = logging.getLogger(__name__)


class BookmarkParser(object):
    """
    References
    http://michaellynn.github.io/2015/10/24/apples-bookmarkdata-exposed/
    """
    HEADER = struct.Struct('< 4s 3I')
    TOC_HEADER = struct.Struct('< I 2H 3I')
    TOC_DATA_HEADER = struct.Struct('< 3I')
    RECORD_HEADER = struct.Struct('< 2I')

    EXPECTED_TYPE_MASK = 0xffffff00
    STRING_TYPE = 0x100
    BYTES_TYPE = 0x200
    NUMBER_TYPE = 0x300
    DATE_TYPE = 0x400
    BOOL_TYPE = 0x500
    ARRAY_TYPE = 0x600
    DICT_TYPE = 0x700  # haven't seen one of these in the wild, yet
    UUID_TYPE = 0x800
    URL_TYPE = 0x900
    NULL_TYPE = 0xA00

    # field_id: ([expected data types], field_name)
    # fields with field_name of None will not be parsed
    FIELDS = {
        0x1003: ([STRING_TYPE, URL_TYPE], 'url_string'),  # string
        0x1004: ([ARRAY_TYPE], 'path'),  # array of path components
        0x1005: ([ARRAY_TYPE], 'inode_path'),  # array of file IDs
        0x1010: ([BYTES_TYPE], 'resource_props'),  # byte array of props - three 8-byte ints: [0] target flags, [1] flag validity, [2] unknown (always 0)
        0x1020: ([STRING_TYPE, URL_TYPE], 'target_filename'),  # string
        0x1030: ([NUMBER_TYPE], 'target_inode'),  # int
        0x1040: ([DATE_TYPE], 'creation_date'),  # date
        0x2000: ([ARRAY_TYPE], 'volume_info_depths'),  # array of depths
        0x2002: ([STRING_TYPE, URL_TYPE], 'volume_path'),  # string
        0x2005: ([STRING_TYPE, URL_TYPE], 'volume_url'),  # CFURL
        0x2010: ([STRING_TYPE], 'volume_name'),  # string
        0x2011: ([STRING_TYPE, UUID_TYPE], 'volume_uuid'),  # string
        0x2012: ([NUMBER_TYPE], 'volume_size'),  # int
        0x2013: ([DATE_TYPE], 'volume_creation_date'),  # date
        0x2020: ([BYTES_TYPE], 'volume_props'),  # byte array of props - three 8-byte ints: [0] target flags, [1] flag validity, [2] unknown (always 0)
        0x2030: ([BOOL_TYPE], 'volume_was_boot'),  # bool (existence indicates True)
        0x2040: ([NUMBER_TYPE], 'disk_image_depth'),  # 32-bit int
        0x2050: ([STRING_TYPE, URL_TYPE], 'volume_mount_point'),  # CFURL
        0xc001: ([NUMBER_TYPE], None),  # 'volume_home_dir_relative_path_component_count' - index of parent in path components
        0xc011: ([STRING_TYPE], 'user_name'),  # string
        0xc012: ([NUMBER_TYPE], 'user_uid'),  # int
        0xd001: ([BOOL_TYPE], None),  # 'wasFileIDFormat' bool
        0xd010: ([NUMBER_TYPE], None),  # 'creation_options' flags for CFURLCreateBookmarkData
        0xe003: ([ARRAY_TYPE], None),   # 'url_length' - array
        0xf017: ([STRING_TYPE], 'display_name'),
        0xf020: ([BYTES_TYPE], None),  # 'effective_icon_data' - icns/icon image byte array
        0xf021: ([BYTES_TYPE], None),  # 'effective_flattened_icon_ref' - img byte array
        0xf022: ([BYTES_TYPE], None),  # 'type_binding_data' - dnib byte array
        0xf030: ([NUMBER_TYPE], 'bookmark_creation_time'),  # 'bookmark_creation_time' exp 64-bit float seconds since 1/1/2001 (e.g. b'68CB95EB545DB841')
        0xf080: ([BYTES_TYPE], 'sandbox_rw_extension'),  # semi-colon separated values string (from byte-array)
        0xf081: ([BYTES_TYPE], 'sandbox_ro_extension'),  # semi-colon separated values string (from byte-array)
        0xfe00: ([BYTES_TYPE], 'alias_data'),  # Alias byte array
        0x800000d0: ([NUMBER_TYPE], None),  # 'nsurl_document_identifier_key' https://developer.apple.com/reference/foundation/nsurldocumentidentifierkey
        0x80000190: ([NUMBER_TYPE], None),  # 'nsurl_document_identifier_key' https://developer.apple.com/reference/foundation/nsurldocumentidentifierkey
        0x80000194: ([NUMBER_TYPE], None),  # 'nsurl_document_identifier_key' https://developer.apple.com/reference/foundation/nsurldocumentidentifierkey
        0x800001ac: ([NUMBER_TYPE], None),  # 'nsurl_document_identifier_key' https://developer.apple.com/reference/foundation/nsurldocumentidentifierkey
        0x800001b0: ([NUMBER_TYPE], None),  # 'nsurl_document_identifier_key' https://developer.apple.com/reference/foundation/nsurldocumentidentifierkey
        0x800001c0: ([NUMBER_TYPE], None),  # 'nsurl_document_identifier_key' https://developer.apple.com/reference/foundation/nsurldocumentidentifierkey
        0x800001d8: ([NUMBER_TYPE], None),  # 'nsurl_document_identifier_key' https://developer.apple.com/reference/foundation/nsurldocumentidentifierkey
        0x80000218: ([NUMBER_TYPE], None),  # 'nsurl_document_identifier_key' https://developer.apple.com/reference/foundation/nsurldocumentidentifierkey
        0x80000278: ([NUMBER_TYPE], None),  # 'nsurl_document_identifier_key' https://developer.apple.com/reference/foundation/nsurldocumentidentifierkey
        0x8000027c: ([NUMBER_TYPE], None),  # 'nsurl_document_identifier_key' https://developer.apple.com/reference/foundation/nsurldocumentidentifierkey
    }

    # https://opensource.apple.com/source/CF/CF-1153.18/CFURLPriv.h.auto.html
    RESOURCE_PROPERTY_FLAGS = [
        (0x00000001, 'IsRegularFile'),
        (0x00000002, 'IsDirectory'),
        (0x00000004, 'IsSymbolicLink'),
        (0x00000008, 'IsVolume'),
        (0x00000010, 'IsPackage'),
        (0x00000020, 'IsSystemImmutable'),
        (0x00000040, 'IsUserImmutable'),
        (0x00000080, 'IsHidden'),
        (0x00000100, 'HasHiddenExtension'),
        (0x00000200, 'IsApplication'),
        (0x00000400, 'IsCompressed'),
        (0x00000800, 'CanSetHiddenExtension'),
        (0x00001000, 'IsReadable'),
        (0x00002000, 'IsWriteable'),
        (0x00004000, 'IsExecutable'),  # execute files or search directories
        (0x00008000, 'IsAliasFile'),
        (0x00010000, 'IsMountTrigger'),
    ]

    # https://opensource.apple.com/source/CF/CF-1153.18/CFURLPriv.h.auto.html
    VOLUME_PROPERTY_FLAGS = [
        (0x1, 'IsLocal'),         # Local device (vs. network device)
        (0x2, 'IsAutomount'),     # Mounted by the automounter
        (0x4, 'DontBrowse'),      # Hidden from user browsing
        (0x8, 'IsReadOnly'),      # Mounted read-only
        (0x10, 'IsQuarantined'),  # Mounted with quarantine bit
        (0x20, 'IsEjectable'),
        (0x40, 'IsRemovable'),
        (0x80, 'IsInternal'),
        (0x100, 'IsExternal'),
        (0x200, 'IsDiskImage'),
        (0x400, 'IsFileVault'),
        (0x800, 'IsLocaliDiskMirror'),
        (0x1000, 'IsiPod'),
        (0x2000, 'IsiDisk'),
        (0x4000, 'IsCD'),
        (0x8000, 'IsDVD'),
        (0x10000, 'IsDeviceFileSystem'),
        (0x100000000, 'SupportsPersistentIDs'),
        (0x200000000, 'SupportsSearchFS'),
        (0x400000000, 'SupportsExchange'),
        (0x1000000000, 'SupportsSymbolicLinks'),
        (0x2000000000, 'SupportsDenyModes'),
        (0x4000000000, 'SupportsCopyFile'),
        (0x8000000000, 'SupportsReadDirAttr'),
        (0x10000000000, 'SupportsJournaling'),
        (0x20000000000, 'SupportsRename'),
        (0x40000000000, 'SupportsFastStatFS'),
        (0x80000000000, 'SupportsCaseSensitiveNames'),
        (0x100000000000, 'SupportsCasePreservedNames'),
        (0x200000000000, 'SupportsFLock'),
        (0x400000000000, 'HasNoRootDirectoryTimes'),
        (0x800000000000, 'SupportsExtendedSecurity'),
        (0x1000000000000, 'Supports2TBFileSize'),
        (0x2000000000000, 'SupportsHardLinks'),
        (0x4000000000000, 'SupportsMandatoryByteRangeLocks'),
        (0x8000000000000, 'SupportsPathFromID'),
        (0x20000000000000, 'IsJournaling'),
        (0x40000000000000, 'SupportsSparseFiles'),
        (0x80000000000000, 'SupportsZeroRuns'),
        (0x100000000000000, 'SupportsVolumeSizes'),
        (0x200000000000000, 'SupportsRemoteEvents'),
        (0x400000000000000, 'SupportsHiddenFiles'),
        (0x800000000000000, 'SupportsDecmpFSCompression'),
        (0x1000000000000000, 'Has64BitObjectIDs')
    ]

    @classmethod
    def get_toc(cls, buf, data_offset):
        toc_offset, = struct.unpack_from('<I', buf, data_offset)
        table_of_contents = []

        toc_count = 0
        while toc_offset > 0:
            toc_info, toc_offset = cls.parse_toc(buf, data_offset + toc_offset, data_offset, toc_count)
            table_of_contents.extend(toc_info)
            toc_count += 1
        return table_of_contents, toc_count

    @classmethod
    def parse_bookmark(cls, fullpath, idx, item_name, buf):
        if buf is None or len(buf) < cls.HEADER.size:
            return []
        magic, _size, _version, data_offset = cls.HEADER.unpack_from(buf)
        if magic != b'book' and magic != b'alis':
            return []
        table_of_contents, toc_count = cls.get_toc(buf, data_offset)

        all_data = [{'bookmark_index': idx} for _i in range(toc_count)]
        embedded = defaultdict(list)
        for toc_entry in table_of_contents:
            cur_toc_entry = all_data[toc_entry['index']]
            if 'toc_depth' not in cur_toc_entry:
                cur_toc_entry['toc_depth'] = toc_entry['depth']
            record_offset = toc_entry['record_offset']
            record_length, record_data_type = cls.RECORD_HEADER.unpack_from(buf, record_offset)
            cls.process_field(fullpath, buf, item_name, data_offset, cur_toc_entry,
                              toc_entry['record_type'], record_offset, record_length, record_data_type)
            if 'alias_data' in cur_toc_entry:
                for alias_record in AliasParser.parse(fullpath, idx, cur_toc_entry.pop('alias_data')):
                    embedded[toc_entry['index']].append(dict(alias_record, bookmark_index=idx))
        # yield embedded alias records immediately following parent bookmark entry record
        for rec_idx, record in enumerate(all_data):
            yield record
            for embedded_record in embedded.get(rec_idx, []):
                yield embedded_record

    @classmethod
    def process_field(cls, fullpath, buf, item_name, data_offset, cur_toc_entry,
                      rec_type, record_offset, record_length, record_data_type):
        if rec_type in cls.FIELDS:
            expected_data_types, field_name = cls.FIELDS[rec_type]
            if field_name is None:  # we know what it is, we just don't want to process it
                return
            general_type = record_data_type & cls.EXPECTED_TYPE_MASK
            if general_type not in expected_data_types + [cls.NULL_TYPE]:  # anything could be NULL
                logger.error(
                    "Unexpected data type {:#x} for record type {:#x} ({}) in file '{}', please report.", record_data_type, rec_type, field_name, fullpath)
                return
            record_data_offset = record_offset + cls.RECORD_HEADER.size
            data = buf[record_data_offset: record_data_offset + record_length]
            field_dict = cls.decode_value(field_name, buf, data_offset, rec_type, record_length, record_data_type, data)
            cls.update_record(cur_toc_entry, field_dict)
        else:
            logger.warning(
                "Unknown bookmark record/data type ({}/{}) in item '{}' from file '{}', please report.", rec_type, record_data_type, item_name, fullpath)

    @staticmethod
    def update_record(record, field_dict, warn=True):
        for k, v in field_dict.items():
            if k in record:
                if warn:
                    logger.error(
                        "Could not update record due to duplicate key in level. Initial value: {}/{}. Attempted update: {}/{}.",
                        k, record[k], k, field_dict[k]
                    )
            else:
                record[k] = v

    @classmethod
    def _parse_record_data_601(cls, data, record_length, buf, data_offset):
        pointer_ct = record_length // 4
        pointers = struct.unpack('<{}I'.format(pointer_ct), data)
        parsed_values = []
        for p in pointers:
            component_offset = p + data_offset
            component_length, component_data_type = cls.RECORD_HEADER.unpack_from(buf, component_offset)
            component_data_offset = component_offset + cls.RECORD_HEADER.size
            component_data = buf[component_data_offset:component_data_offset + component_length]
            parsed_values.append(
                cls.parse_record_data(buf, data_offset, component_length,
                                      component_data_type, component_data))
        return parsed_values

    @classmethod
    def _parse_record_data_902(cls, data, record_length, buf, data_offset):
        parsed = cls._parse_record_data_601(data, record_length, buf, data_offset)
        rec_count = len(parsed)
        if rec_count == 2:
            return urljoin(parsed[0], parsed[1])
        joined = '/'.join(parsed)
        logger.warning("Unexpected record count {} in URL array (expected 2): '{}', please report.", rec_count, joined)
        return joined

    @classmethod
    def _parse_record_data_a01(cls, _data, record_length, *args):
        if record_length != 0:
            logger.warning("Unexpected data length {} in bookmark data type 0xA01, please report.", record_length)
        return None

    @classmethod
    def parse_record_data(cls, buf, data_offset, record_length, data_type, data):
        identity = lambda x, *args: x
        parsers = {
            0x101: lambda x, *args: x.decode('utf-8'),             # UTF-8 String
            0x201: identity,                                       # byte array, up to caller to handle
            # 0x300 CFNumberType https://developer.apple.com/reference/corefoundation/cfnumbertype
            0x301: lambda x, *args: struct.unpack('<b', data)[0],  # sInt8Type: Eight-bit, signed integer. The SInt8 data type is defined in MacTypes.h
            0x302: lambda x, *args: struct.unpack('<h', data)[0],  # sInt16Type: Sixteen-bit, signed integer. The SInt16 data type is defined in MacTypes.h
            0x303: lambda x, *args: struct.unpack('<i', data)[0],  # sInt32Type: Thirty-two-bit, signed integer. The SInt32 data type is defined in MacTypes.h
            0x304: lambda x, *args: struct.unpack('<q', data)[0],  # sInt64Type: Sixty-four-bit, signed integer. The SInt64 data type is defined in MacTypes.h
            0x305: lambda x, *args: struct.unpack('<f', data)[0],  # float32Type: Thirty-two-bit real. The Float32 data type is defined in MacTypes.h
            0x306: lambda x, *args: struct.unpack('<d', data)[0],  # float64Type: Sixty-four-bit real. The Float64 data type is defined in MacTypes.h and conforms to the 64-bit IEEE 754 standard
            0x307: lambda x, *args: struct.unpack('<B', data)[0],  # charType: Basic C char type
            0x308: lambda x, *args: struct.unpack('<H', data)[0],  # shortType: Basic C short type
            0x309: lambda x, *args: struct.unpack('<I', data)[0],  # intType: Basic C int type
            0x30A: lambda x, *args: struct.unpack('<L', data)[0],  # longType: Basic C long type
            0x30B: lambda x, *args: struct.unpack('<Q', data)[0],  # longLongType: Basic C long long type
            0x30C: lambda x, *args: struct.unpack('<f', data)[0],  # floatType: Basic C float type
            0x30D: lambda x, *args: struct.unpack('<d', data)[0],  # doubleType: Basic C double type
            0x30E: lambda x, *args: struct.unpack('<I', data)[0],  # cfIndexType: CFIndex value, see https://developer.apple.com/reference/corefoundation/cfindex
            0x30F: lambda x, *args: struct.unpack('<I', data)[0],  # nsIntegerType: NSInteger value, see https://developer.apple.com/reference/objectivec/nsinteger
            # 0x310 cgFloatType: Apparently this can be either 32-bit or 64-bit float,
            # dependent on target build architecture; CGFloat value, see https://developer.apple.com/reference/coregraphics/cgfloat
            0x400: lambda x, *args: parse_mac_absolute_time(struct.unpack_from('>d', data)[0]),  # Timestamp
            0x500: lambda x, *args: False,                                          # bool False if exists
            0x501: lambda x, *args: True,                                           # bool True if exists
            0x601: cls._parse_record_data_601,                     # array of pointers to data within bookmark (32-bit int + data_offset)
            0x801: lambda x, * args: guid_str_from_bytes(data, 'be'),  # UUID raw bytes
            0x901: lambda x, *args: x.decode('utf-8'),             # CFURL (UTF-8 String)
            0x902: cls._parse_record_data_902,                     # CFURL via array of pointers (multi-part URL)
            0xA01: cls._parse_record_data_a01,                     # CFNull
        }

        return parsers.get(data_type, identity)(data, record_length, buf, data_offset)

    @classmethod
    def _decode_sandbox_value(cls, parsed):
        parts = parsed.split(b';')
        return {
            'sandbox_uuid': parts[0].decode('utf-8'),
            'sandbox_path': parts[-1].rstrip(b'\x00').decode('utf-8'),
        }

    @classmethod
    def decode_value(cls, field_name, buf, data_offset, item_type, record_length, data_type, data):
        if field_name is None:
            return {}
        parsed = cls.parse_record_data(buf, data_offset, record_length, data_type, data)

        decoders = {
            0x1004: lambda x: {field_name: cls.join_path(x)},
            0x1005: lambda x: {field_name: cls.join_path(x)},
            0x1010: lambda x: {field_name: interpret_flags(struct.unpack_from('<Q', x[:8])[0], cls.RESOURCE_PROPERTY_FLAGS)},
            0x2000: lambda x: {field_name: ', '.join([str(y) for y in x])},
            0x2020: lambda x: {field_name: interpret_flags(struct.unpack_from('<Q', x[:8])[0], cls.VOLUME_PROPERTY_FLAGS)},
            0xf030: lambda x: {field_name: parse_mac_absolute_time(x)},
            0xf080: cls._decode_sandbox_value,
            0xf081: cls._decode_sandbox_value
        }
        return decoders.get(item_type, lambda x: {field_name: x})(parsed)

    @classmethod
    def join_path(cls, array):
        return '/' + '/'.join([str(x) for x in array if x])

    @classmethod
    def parse_toc(cls, buf, offset, data_offset, toc_index):
        contents = []
        _data_length, record_type, flags, depth, next_toc, count = cls.TOC_HEADER.unpack_from(buf, offset)
        for i in range(count):
            (record_type,
             record_offset,
             flags) = cls.TOC_DATA_HEADER.unpack_from(buf,
                                                      offset + cls.TOC_HEADER.size + i * cls.TOC_DATA_HEADER.size)

            contents.append({'record_type': record_type,
                             'record_type_hex': hex(record_type),
                             'flags': flags,
                             'record_offset': record_offset + data_offset,
                             'hex_offset': hex(record_offset + data_offset),
                             'index': toc_index,
                             'depth': depth})

        return contents, next_toc
